File: Excel/700-799/764/764 Challenge.py
import operator as op
import pandas as pd

path = "700-799/764/764 Reverse Polish Notation_2.xlsx"
input = pd.read_excel(path, usecols="A", nrows=10)
test = pd.read_excel(path, usecols="B", nrows=10)

def eval_rpn(tokens):
    stack = []
    for t in tokens:
        if t in '+-*/':
            if len(stack) > 2:
                res = stack[0]
                for n in stack[1:]:
                    res = eval(f"{res}{t}{n}")
                stack = [res]
            else:
                b = stack.pop()
                a = stack.pop()
                stack.append(eval(f"{a}{t}{b}"))
        else:
            stack.append(int(t))
    return stack[0]

input['Result'] = input['Text'].apply(lambda x: eval_rpn(x.split(", "))).astype(int)

print(input["Result"].equals(test["Expected Answer"])) # True

File: Excel/700-799/764/764 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/700-799/764/764 Reverse Polish Notation_2.xlsx"
input = read_excel(path, range = "A1:A10")
test  = read_excel(path, range = "B1:B10")

eval_rpn = function(tokens) {
  stack = c()
  for (t in tokens) {
    if (t %in% c('+', '-', '*', '/')) {
      if (length(stack) > 2) {
        res = stack[1]
        for (n in stack[-1]) {
          res = eval(parse(text = paste0(res, t, n)))
        }
        stack = res
      } else {
        b = tail(stack, 1)
        a = tail(stack, 2)[1]
        stack = c(head(stack, -2), eval(parse(text = paste0(a, t, b))))
      }
    } else {
      stack = c(stack, as.numeric(t))
    }
  }
  stack[1]
}

result = input %>%
  mutate(Result = map_dbl(Text, ~eval_rpn(strsplit(.x, ", ")[[1]])))

all.equal(result$Result, test$`Expected Answer`) 
# [1] TRUE
